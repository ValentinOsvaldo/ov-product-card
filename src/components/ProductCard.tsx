import React, { createContext } from 'react';
import { useProduct } from '../hooks';
import {
  onChangeArgs,
  Product,
  ProductContextProps,
  InitialValues,
  ProductCardHandlers,
} from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';

export interface Props {
  // children?: React.ReactElement | React.ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  product: Product;
  style?: React.CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const { counter, increaseBy, maxCount, isMaxCountReached, reset } =
    useProduct({
      onChange,
      product,
      value,
      initialValues,
    });

  return (
    <Provider
      value={{
        product,
        counter,
        increaseBy,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className || ''}`} style={style}>
        {children({
          count: counter,
          maxCount,
          product,
          isMaxCountReached,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
