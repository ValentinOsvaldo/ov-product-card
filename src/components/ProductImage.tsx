import React, { useContext } from 'react';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface Props {
  className?: string;
  img?: string;
  style?: React.CSSProperties;
}

export const ProductImage = ({ img, className, style }: Props) => {
  const { product } = useContext(ProductContext);
  const imgToShow = img ? img : product.img ?? noImage;

  return (
    <img
      src={imgToShow}
      alt="Product Image"
      className={`${styles.productImg} ${className || ''}`}
      style={style}
    />
  );
};
