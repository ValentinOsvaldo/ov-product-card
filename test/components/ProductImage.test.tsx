import React from "react";
import renderer from 'react-test-renderer'
import { ProductImage, ProductCard } from '../../src/components';
import {  product2 } from '../data/products';

describe('Product Image', () => {
  test('Debe mostrar el componente correctamente con la imagen personalizada', () => {
    const wrapper = renderer.create(
      <ProductImage img="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png" />
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  
  test('Debe mostrar el componente correctamente con la imagen del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {
          () => (
            <ProductImage />
          )
        }
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
