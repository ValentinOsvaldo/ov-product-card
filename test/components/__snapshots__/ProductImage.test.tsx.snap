// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Image Debe mostrar el componente correctamente con la imagen del producto 1`] = `
<div
  className="productCard "
>
  <img
    alt="Product Image"
    className="productImg "
    src="./coffe-mug.png"
  />
</div>
`;

exports[`Product Image Debe mostrar el componente correctamente con la imagen personalizada 1`] = `
<img
  alt="Product Image"
  className="productImg "
  src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png"
/>
`;
